services/mac_manager/src/main/java/com/futurewei/alcor/macmanager/pool/impl/BitmapPoolImpl.java

```diff
@@ -0,0 +1,261 @@
+                    }else{
+                        // if not put success, it already allocated
+                        macRangePartition.incUsed();
+                        needUpdate = true;
+                    }
+                }
+
+                if (needUpdate) {
+                    macRangePartitionRepository.addItem(macRangePartition);
+                }
+
+            } finally {
+                try {
+                    lock.unlock(id);
+                } catch (DistributedLockException e) {
+                    logger.error("unlock cluster lock {} failed: {}", id, e.getMessage());
+                }
+            }
+        }
+        throw new MacAddressRetryLimitExceedException(MacManagerConstant.MAC_EXCEPTION_RETRY_LIMIT_EXCEED);
+    }
+
+    @Override
+    public Set<String> allocateBulk(String oui, MacRange macRange, int size) throws MacAddressFullException,
+            MacAddressRetryLimitExceedException, CacheException{
+        Set<String> newAllocatedMacs = new HashSet<>();
+        String rangeId = macRange.getRangeId();
+        long used = macRangeMappingRepository.size(rangeId);
+        if(used + size >= macRange.getCapacity()){
+            throw new MacAddressFullException(MacManagerConstant.MAC_EXCEPTION_MACADDRESS_FULL);
+        }
+
+        int partitionNums = (int) (macRange.getCapacity()/PARTITION_SIZE + 1);
+        int partitionIndex = 0;
+        int retryTime = 0;
+        while (retryTime < nRetryLimit) {
+
+            boolean locked = false;
+            while (!locked) {
+                partitionIndex = RandomUtils.nextInt(0, partitionNums);
+                locked = lock.tryLock(rangeId + "_" + partitionIndex);
+            }
+
+            String id = rangeId + "_" + partitionIndex;
+            try {
+                boolean needUpdate = false;
+                MacRangePartition macRangePartition = macRangePartitionRepository.findItem(id);
+
+                // if null create a new one
+                if (macRangePartition == null) {
+                    macRangePartition = new MacRangePartition(id, partitionIndex, PARTITION_SIZE * partitionIndex,
+                            PARTITION_SIZE * (partitionIndex + 1));
+                    needUpdate = true;
+                }
+
+                while (macRangePartition.getUsed() < macRangePartition.getTotal()) {
+                    long macLong = generate(macRangePartition);
+                    if(check(rangeId, macLong)) {
+                        macRangePartition.incUsed();
+                        needUpdate = true;
+                        newAllocatedMacs.add(longToMac(oui, macLong));
+                        if(newAllocatedMacs.size() >= size){
+                            macRangePartitionRepository.addItem(macRangePartition);
+                            return newAllocatedMacs;
+                        }
+                    }else{
+                        // if not put success, it already allocated
```

Comment:
If put is not successful, we shouldn't increase the Used value, right?

Reply:
If the MAC address fails to be added to the MAC address table, the MAC address is in use.